opslevel/datasource_opslevel_tier.go

```diff
@@ -1,94 +1,137 @@
+			},
+			"name": schema.StringAttribute{
+				Description: "The name of the domain.",
+				Computed:    true,
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"filter": getDatasourceFilter(validFieldNames),
+		},
+	}
+}
+
+func (d *TierDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data TierDataSourceModel
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tiers, err := d.client.ListTiers()
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read tier datasource, got error: %s", err))
+		return
+	}
+
+	tier, err := filterTiers(tiers, data.Filter)
+	if err != nil {
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to filter tier datasource, got error: %s", err))
+		return
+	}
+
+	tierDataModel := NewTierDataSourceModel(ctx, *tier, data.Filter)
+
+	// Save data into Terraform state
+	tflog.Trace(ctx, "read an OpsLevel Tier data source")
+	// resp.Diagnostics.Append(diags...)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &tierDataModel)...)
+}
 
 // func filterTiers(levels []opslevel.Tier, field string, value string) (*opslevel.Tier, error) {
-// 	if value == "" {
-// 		return nil, fmt.Errorf("Please provide a non-empty value for filter's value")
-// 	}
-
-// 	var output opslevel.Tier
-// 	found := false
-// 	for _, item := range levels {
-// 		switch field {
-// 		case "alias":
-// 			if item.Alias == value {
-// 				output = item
-// 				found = true
-// 			}
-// 		case "id":
-// 			if string(item.Id) == value {
-// 				output = item
-// 				found = true
-// 			}
-// 		case "index":
-// 			if v, err := strconv.Atoi(value); err == nil && item.Index == v {
-// 				output = item
-// 				found = true
-// 			}
-// 		case "name":
-// 			if item.Name == value {
-// 				output = item
-// 				found = true
-// 			}
-// 		}
-// 		if found {
-// 			break
-// 		}
-// 	}
-
-// 	if !found {
-// 		return nil, fmt.Errorf("Unable to find tier with: %s==%s", field, value)
-// 	}
-// 	return &output, nil
-// }
-
-// func datasourceTierRead(d *schema.ResourceData, client *opslevel.Client) error {
-// 	results, err := client.ListTiers()
-// 	if err != nil {
-// 		return err
-// 	}
-
-// 	field := d.Get("filter.0.field").(string)
-// 	value := d.Get("filter.0.value").(string)
-
-// 	item, itemErr := filterTiers(results, field, value)
-// 	if itemErr != nil {
-// 		return itemErr
-// 	}
-
-// 	d.SetId(string(item.Id))
-// 	d.Set("alias", item.Alias)
-// 	d.Set("index", item.Index)
-// 	d.Set("name", item.Name)
-
-// 	return nil
-// }
+func filterTiers(tiers []opslevel.Tier, filter FilterModel) (*opslevel.Tier, error) {
+	if filter.Value.Equal(types.StringValue("")) {
+		return nil, fmt.Errorf("Please provide a non-empty value for filter's value")
+	}
+	for _, tier := range tiers {
+		switch filter.Field.ValueString() {
```

Comment:
Should this do anything in the default case?

Reply:
No need for a default case here.
Assuming the filter looks like:
```terraform
filter {
  field = "alias"
  value = "tier_1"
}
```
We want to check all tiers for the alias `tier_1`.
If this is the last item in `tiers`, we don't need to log anything along the way. If nothing is found after all tiers are looked through then the error below catches that case.